frontend/packages/kubevirt-plugin/src/components/vms/pending-changes-warning.tsx

```diff
@@ -0,0 +1,200 @@
+    [VMTabEnum.env]: [],
+    [VMTabEnum.nics]: [],
+    [VMTabEnum.disks]: [],
+  };
+
+  Object.keys(pendingChanges).forEach((key) => {
+    const pc = pendingChanges[key];
+    pc.isPendingChange && pendingChangesByTab[pc.vmTab].push(key);
+  });
+
+  return pendingChangesByTab;
+};
+
+type WarningTabRowProps = {
+  tabName: string;
+  tabProps: string[];
+  pendingChanges: PendingChanges;
+};
+
+const WarningTabRow: React.FC<WarningTabRowProps> = ({ tabName, tabProps, pendingChanges }) => (
+  <Breadcrumb className="kv-warning--pf-c-breadcrumb">
+    <BreadcrumbHeading key={`${tabName}-header-${tabProps.join('-')}`}>{tabName}</BreadcrumbHeading>
+    <BreadcrumbItem key={`${tabName}-${tabProps.join('-')}`}>
+      {tabProps.map((key, idx) => (
+        <div key={`${tabName}-${key}`} className="kv-warning-changed-attrs">
+          <Button isInline onClick={pendingChanges[key].execAction} variant={ButtonVariant.link}>
+            {key}
+          </Button>
+          <div className="kv-warning--comma-margin">{idx === tabProps.length - 1 ? '' : ','}</div>
+        </div>
+      ))}
+    </BreadcrumbItem>
+  </Breadcrumb>
+);
+
+export const getPendingChanges = (
+  vm: VMKind,
+  vmWrapper: VMWrapper,
+  vmiWrapper: VMIWrapper,
+): PendingChanges => ({
+  [IsPendingChange.flavor]: {
+    isPendingChange: isFlavorChanged(vmWrapper, vmiWrapper),
+    execAction: () => vmFlavorModal({ vmLike: vm, blocking: true }),
+    vmTab: VMTabEnum.details,
+  },
+  [IsPendingChange.cdroms]: {
+    isPendingChange: isCDROMChanged(vmWrapper, vmiWrapper),
+    execAction: () => VMCDRomModal({ vmLikeEntity: vm, modalClassName: 'modal-lg' }),
+    vmTab: VMTabEnum.details,
+  },
+  [IsPendingChange.bootOrder]: {
+    isPendingChange: isBootOrderChanged(vmWrapper, vmiWrapper),
+    execAction: () => BootOrderModal({ vmLikeEntity: vm, modalClassName: 'modal-lg' }),
+    vmTab: VMTabEnum.details,
+  },
+  [IsPendingChange.env]: {
+    isPendingChange: isEnvDisksChanged(vmWrapper, vmiWrapper),
+    execAction: () =>
+      history.push(`/ns/${getNamespace(vm)}/virtualmachines/${getName(vm)}/environment`),
+    vmTab: VMTabEnum.env,
+  },
+  [IsPendingChange.nics]: {
+    isPendingChange: isNicsChanged(vmWrapper, vmiWrapper),
+    execAction: () => history.push(`/ns/${getNamespace(vm)}/virtualmachines/${getName(vm)}/nics`),
+    vmTab: VMTabEnum.nics,
+  },
+  [IsPendingChange.disks]: {
+    isPendingChange: isDisksChanged(vmWrapper, vmiWrapper),
+    execAction: () => history.push(`/ns/${getNamespace(vm)}/virtualmachines/${getName(vm)}/disks`),
+    vmTab: VMTabEnum.disks,
+  },
+});
+
+export const PendingChangesWarning: React.FC<PendingChangesWarningProps> = ({
+  vmLikeEntity,
+  vmi: vmiProp,
+}) => {
+  const vm: VMKind = asVM(getLoadedData(vmLikeEntity));
+  const vmi = getLoadedData(vmiProp);
+
+  const vmWrapper = new VMWrapper(vm);
+  const vmiWrapper = new VMIWrapper(vmi);
+
+  const [pendingChanges, setPendingChanges] = React.useState<PendingChanges>(
```

Comment:
does having a state makes sense here?
the useEffect will fire every time with current firehose configuration.

does calling getPendingChanges make the ui slow? if yes than we shoul think of some other mechanism of optimizing it

Reply:
how can I measure if the UI acting slow due to the getPendingChanges calls?
and if it is, what other mechanism you propose ?

Reply:
you can just try this https://stackoverflow.com/a/46649775 but for rendering basic console.log("render") might suffice to see if there is a very big ammount of renderings happening.

btw the speed also depends on your backend resources and how often the update